src/queue/multilevel.rs

```diff
@@ -0,0 +1,885 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! A multilevel feedback task queue. Long-running tasks are pushed to levels
+//! with lower priority.
+//!
+//! The task queue requires [`TaskCell`] to contain [`MultilevelQueueExtras`]
+//! as its extras. Additionally, the accompanying [`MultilevelRunner`] must be
+//! used to collect necessary information.
+
+use super::{LocalQueue, Pop, TaskCell, TaskInjector};
+use crate::runner::{LocalSpawn, Runner, RunnerBuilder};
+
+use crossbeam_deque::{Injector, Steal, Stealer, Worker};
+use dashmap::DashMap;
+use init_with::InitWith;
+use rand::prelude::*;
+use std::cell::Cell;
+use std::cmp;
+use std::iter;
+use std::sync::atomic::{AtomicBool, AtomicU32, AtomicU64, AtomicUsize, Ordering::SeqCst};
+use std::sync::{Arc, RwLock};
+use std::time::{Duration, Instant};
+
+const LEVEL_NUM: usize = 3;
+
+/// The chance ratio of level 1 and level 2 tasks.
+const CHANCE_RATIO: u32 = 4;
+
+const DEFAULT_CLEANUP_OLD_MAP_INTERVAL: Duration = Duration::from_secs(10);
+
+/// When total elapsed time exceeds this value, it will try to adjust level
+/// chances and reset the total elapsed time.
+const ADJUST_CHANCE_INTERVAL: Duration = Duration::from_secs(1);
+
+/// When the deviation between the target and the actual level 0 proportion
+/// exceeds this value, level chances need to be adjusted.
+const ADJUST_CHANCE_THRESHOLD: f64 = 0.05;
+
+// The chance values below are the numerators of fractions with u32::max_value()
+// as the denominator.
+const INIT_LEVEL0_CHANCE: u32 = 3_435_973_836; // 0.8
+const MIN_LEVEL0_CHANCE: u32 = 1 << 31; // 0.5
+const MAX_LEVEL0_CHANCE: u32 = 4_209_067_949; // 0.98
+const ADJUST_AMOUNT: u32 = (MAX_LEVEL0_CHANCE - MIN_LEVEL0_CHANCE) / 8; // 0.06
+
+/// The injector of a multilevel task queue.
+pub struct MultilevelQueueInjector<T> {
+    level_injectors: Arc<[Injector<T>; LEVEL_NUM]>,
+    manager: Arc<LevelManager>,
+}
+
+impl<T> Clone for MultilevelQueueInjector<T> {
+    fn clone(&self) -> Self {
+        Self {
+            level_injectors: self.level_injectors.clone(),
+            manager: self.manager.clone(),
+        }
+    }
+}
+
+/// The extras for the task cells pushed into a multilevel task queue.
+#[derive(Debug, Clone)]
+pub struct MultilevelQueueExtras {
+    /// The identifier of the task.
+    task_id: u64,
+    /// The instant when the task cell is pushed to the queue.
+    schedule_time: Option<Instant>,
+    /// The time spent on handling this task.
+    running_time: Option<Arc<ElapsedTime>>,
```

Comment:
Why use `Option`? Is `ElapsedTime::zero()` sufficient enough?

Reply:
`running_time` is created and returned by `TaskElapsedMap`. So we don't need the extra allocation caused by `Arc` when initializing `MultilevelQueueExtras`.